### ISSUE TITLE
debugging theta['dm'] always converging to zero

### ISSUE THREAD

**Issue by jsalfity** [jsalfity/sensitivity_games#9]
Running an experiment with `traj_cost.addThetaCost('dm', Quadratic(n=0, d=0, weight=0.1))` is still resulting in `dm` converging to zero. Below is the code to insert `weight=0.1`.
https://github.com/jsalfity/sensitivity_games/blob/2914adb3e5460203a30f1482f32c7e397ec7b718/sensitivity_games/experiments/point_mass_regulation_train.py#L59

Re-evaluating the cost function calculation, I just want to double check a few things are correct:

- Our method to achieve gradient ascent:
https://github.com/jsalfity/sensitivity_games/blob/2914adb3e5460203a30f1482f32c7e397ec7b718/sensitivity_games/experiments/point_mass_regulation_train.py#L64

- Our method to evaluate the trajectory cost, paying close attention to the negative sign for `theta`
https://github.com/jsalfity/sensitivity_games/blob/2914adb3e5460203a30f1482f32c7e397ec7b718/sensitivity_games/trajectory_cost.py#L26-L39

- In the `Quadratic.evaluate()` function, the nominal value, `n`, for `theta['dm']` is zero.
https://github.com/jsalfity/sensitivity_games/blob/2914adb3e5460203a30f1482f32c7e397ec7b718/sensitivity_games/quadratic.py#L19-L27

I'm having trouble stepping through the logic that the combination of these three, along with others, are pushing `dm` to go to 'high disturbance' value. My thought process is ... we want our cost function to be smaller, except for `dm` to be larger, which justifies bullet 1 and 2. I'm can't rationalize why bullet 3 is correct, do we really want a `theta[dm]` nominal=0 ?

We wrote some math out , making sure we are still solving this same problem:
<img width="696" alt="Screen Shot 2021-07-28 at 12 08 16 PM" src="https://user-images.githubusercontent.com/20803806/127381651-29036e94-70a8-45ac-a677-8cd67e0ec69c.png">

Another option I was thinking was to add friction or wind resistance, but then of course it wouldn't be a linear problem.

**Comment by dfridovi** [comment]
So we're not solving the problem in that pdf exactly. We're solving the zero sum game with objectives `(J, -J)` rather than `(||\nabla J||, -||\nabla J||)`. 

The three bullets look correct to me (except noted below) but I could be missing something. A couple things to try:
- decrease the weight on the quadratic penalty for `dm` even further
- try initializing `dm` to something strictly positive or strictly negative (maybe `0` is a bad corner case)
- try setting the nominal `dm` value to something nonzero and see what happens. if `dm -> nominal` still, then that's at least consistent but it probably suggests something is really going on here that we're missing.
I do think that seeing some movement on `dm` is pretty critical to confirming that this has a prayer of doing something sensible in a more complicated example.

One bug I do see is that in accumulating the trajectory cost here
https://github.com/jsalfity/sensitivity_games/blob/2914adb3e5460203a30f1482f32c7e397ec7b718/sensitivity_games/trajectory_cost.py#L36
it appears that we're just repeatedly evaluating the same `dm` cost at every time, even though it's time-invariant. That should instead go outside the loop so it only happens once. This could be scaling it large enough that it still drives `dm -> 0` even with `weight = 0.1`.

**Comment by jsalfity** [comment]
oh nice catch! just fixed and pushed to master. 

- I'll run different weights with the traj_cost fix and see outcome.

- I'm not sure what you mean here: "try initializing dm to something strictly positive or strictly negative (maybe 0 is a bad corner case)". It's currently initialized as one:
https://github.com/jsalfity/sensitivity_games/blob/4c15bd5b8a97ed8b5d5b014eb9383aa8acc8e208/sensitivity_games/experiments/point_mass_regulation_train.py#L41-L45

- I'll try different settings for `n`, my intuition says if there isn't a ceiling (implemented as high cost like we discussed) for control effort or resistance through friction or air, `dm` will still go to zero. ahh, this makes my mind hurt, I know we discussed, but still having trouble lining up.

**Comment by jsalfity** [comment]
Ran with different `n` and `weight`. Regardless of large weights, 1e4, and small weights 1e-4, `dm` is converging to the `n`. The large weight is particularly confusing me, because it would seem like the larger the `dm` quadratic cost, the larger the negative sum, and therefore the `total` would be driven smaller.

So it's interesting to see the trend that `dm` is going to `n`, which I would expect if total is composed of the positive sum, rather than negative. Hmm.. still have some digging to do.

**Comment by dfridovi** [comment]
Can you try just setting the `dm` cost weight to zero? What _should_ happen is that `dm` should realize it can blow up really big to make the point mass stay in place. If that doesn't happen, something is very off.

**Comment by jsalfity** [comment]
Good thinking. With `weight=0` and any value of `n`, the `theta['dm']` value isn't moving at all. i.e., it's staying at the initialized value, 1 in this case.

In the case of `dm` blowing up, wouldn't it then be a battle between the state cost and control cost ? state cost would drive the mass to xf, while control cost would drive it to not move (because control effort is too high) ?

**Comment by dfridovi** [comment]
Ok so check if the gradients `dJdtheta` are zero. If so, then that means somehow we're setting things up wrong in pytorch. If not, then it probably means that trick of setting stepsize negative doesn't work.

**Comment by dfridovi** [comment]
As for the battle between state/control costs, that would only affect `K`. `dm` doesn't care about control cost, after all.

**Comment by jsalfity** [comment]
created a new branch `fix_issue9`. I think I'm getting closer to solving this bug

When creating a new `self.B` tensor, we're losing `self.theta['dm']` on the torch computation graph. I'm confirming this by `self.B.grad_fn=None` and `self.B.is_leaf=True`. In the training script, when calling `total_cost.backward()`, the `dynamics.B` has a gradient ! which I don't think it should , it should keep propogating back to reach `theta['dm'].  

https://github.com/jsalfity/sensitivity_games/blob/4d6241088c5a3c8c73c8cc14d60b7bb9c1cbfa4e/sensitivity_games/point_mass.py#L83-L104

**Comment by jsalfity** [comment]
After reading some forums, https://discuss.pytorch.org/t/gradient-of-output-is-0/128449 and https://discuss.pytorch.org/t/best-way-to-convert-a-list-to-a-tensor/59949/6, I found a proposed solution that is so crazy is just might work ---

https://github.com/jsalfity/sensitivity_games/blob/a2bff2ad8159941e04e7ff4a970b83a394eb9015/sensitivity_games/point_mass.py#L96-L102

Looks like `theta['dm']` is increasing, however, at a much slower rate than I would have thought. With the `theta_cost` commented out, https://github.com/jsalfity/sensitivity_games/blob/b7034ea75a62a1ed6a47ee971ee1ebb3711cd8fd/sensitivity_games/trajectory_cost.py#L30-L33
output shows:

```sh
EPOCH: 24999
total_cost: 1237.44873046875
controller.K: tensor([[ 7.2696e-01,  2.3914e+00,  2.8006e-04, -4.6699e-04],
        [ 1.3048e-03, -7.9423e-04,  7.2608e-01,  2.3918e+00]],
       requires_grad=True)
controller.K.grad: tensor([[ 0.7087, -0.0653,  0.5604, -0.0514],
        [ 0.3033, -0.0288,  0.2375, -0.0225]])
dynamics.theta['dm']: tensor([17.5478], requires_grad=True)
dynamics.theta['dm'].grad: tensor([29.3603])
x0: tensor([-3.5986,  0.0000, -2.8772,  0.0000])
xf (goal): tensor([0, 0, 0, 0])
xf (actual): tensor([-0.0748,  0.3870, -0.0603,  0.3094], grad_fn=<AddBackward0>)
```

Started my own PyTorch thread, I wonder if anyone will reply https://discuss.pytorch.org/t/placing-tensor-in-new-tensor-list/128479



With the `theta` cost baked into `total_cost`, the output looks like expected. The `weight` on `addThetaCost` is a contributor towards final `theta['dm']` values.  
After 25000 epochs:

`traj_cost.addThetaCost('dm', Quadratic(n=0, d=0, weight=50))` ---> `dynamics.theta['dm']: tensor([0.83], requires_grad=True)`

`traj_cost.addThetaCost('dm', Quadratic(n=0, d=0, weight=1))` ---> `dynamics.theta['dm']: tensor([11.91], requires_grad=True)`

`traj_cost.addThetaCost('dm', Quadratic(n=0, d=0, weight=0.01))` ---> `dynamics.theta['dm']: tensor([17.22], requires_grad=True)`

**Comment by dfridovi** [comment]
This is beautiful - a little tacky and not sure why it didn't work before, but great that it works now. I'm glad the unit test caught this ;)

**Comment by jsalfity** [comment]
yea, apparently making tensors lists from tensors isn't trivial. 🤷‍♂️🤷‍♂️🤷‍♂️

But now, the `theta['dm'].grad` doesn't seem to settle, because its always decreasing with low weight or increasing with high weight. Seems like that'd be expected, but the unit test isn't going to pass.

https://github.com/jsalfity/sensitivity_games/blob/6b534bb7a0be1129f381f38bff26ff18fbece206/sensitivity_games/tests/test_point_mass.py#L34-L40

**Comment by dfridovi** [comment]
The gradients should always be going to zero _if the algorithm is converging._ So, if they're not going to zero in some conditions it's likely an indicator that you're taking too big steps or something like that and `theta` itself is diverging.

**Comment by jsalfity** [comment]
I tried a few different optimizer settings, but I'll have to dig into torch much more to understand what I'm really after.

I'm wondering if the random initial condition `x0` is screwing with the convergence of `theta['dm']`. Holding `x0` constant throughout the training seems to work a bit better and with higher weight, (>15), i.e. seems like more consistent gradient convergence. More to come..
